### ISSUE TITLE
Creation mutation missing internal fields for onCreate subscription to work

### ISSUE THREAD

**Issue by gsans** [aws/aws-appsync-community#127]
**Issue**
When running the mutation below on the AppSync Console

```
mutation createRestaurant {
  createRestaurant(input: {
    name: "Nobu"
    description: "Great Sushi"
    city: "New York"
  }) {
    id name description city
  }
}
```

On the client complaints missing internal fields while using a creation subscription:
- "Cannot return null for non-nullable type: 'AWSDateTime' within parent 'Restaurant' (/onCreateRestaurant/createdAt)"
- "Cannot return null for non-nullable type: 'AWSDateTime' within parent 'Restaurant' (/onCreateRestaurant/updatedAt)"

The issue seems to be that type on the project is defined as
```
type Restaurant @model {
  id: ID!
  name: String!
  description: String!
  city: String!
}
```
But on the AppSync Console appears as
```
type Restaurant {
	id: ID!
	name: String!
	description: String!
	city: String!
	createdAt: AWSDateTime!
	updatedAt: AWSDateTime!
}
```

**Fix**
You can fix this by adding the missing fields but this is quite unexpected as these are internal fields.

```
mutation createRestaurant {
  createRestaurant(input: {
    name: "Nobu"
    description: "Great Sushi"
    city: "New York"
  }) {
    id name description city createdAt updatedAt
  }
}
```

**Comment by adilusman51** [comment]
@gsans in the example above `createdAt` and `updatedAt` are non-nullable due to appending `!` to the type `AWSDateTime ` , you can remove `!` at the end of these types to make them nullable, then appsync will not return error

**Comment by gsans** [comment]
Sure but these fields were added by AppSync in the first place therefore the issue is that a normal user would be confused that the default setting for subscriptions needs extra fields or otherwise throws an error. Subscriptions should work without any changes

**Comment by buggy** [comment]
@gsans With AppSync subscriptions only receive the fields requested by the mutation. The two relevant sections from the AWS docs are:

> if you’re using MQTT over WebSockets, the client triggering the mutation specifies the selection set that subscribers receive

and

> If you use the pure WebSockets client, selection set filtering is done per client, as each client can define its own selection set. In this case, the subscription selection set must be a subset of the mutation selection set.

See https://docs.aws.amazon.com/appsync/latest/devguide/real-time-data.html

Without requesting `createdAt` and `updatedAt` when calling `createRestaurant` they won't be sent to subscribers. It looks like the client is treating that as a `null` value which is invalid because the field is defined as non-null.

There is already an issue #89 requesting this be fixed.

**Comment by gsans** [comment]
I think you are missing the point here. Besides the technical side of this issue what I find that needs to be fixed is that AppSync default setup leads to a difficult error to troubleshoot in subscriptions. 

My schema didn't have those fields. I may never heard of these fields. But somehow if I try to use a subscription it throws an error. Not because I made an error but because the setup that AppSync used lead me to it. The sensible thing to do for AppSync is to use whatever setup that doesn't lead to an error difficult to explain for the user.

There are few solutions available. One mentioned was removing the mandatory requirement which seems right as these fields are internal. Another option could be changing the subscription behaviour. I don't mind what solution as long as the user has not issues while trying to use the subscription that was setup by AppSync.

**Comment by buggy** [comment]
> Sure but these fields were added by AppSync in the first place therefore the issue is that a normal user would be confused that the default setting for subscriptions needs extra fields or otherwise throws an error.

@gsans This is not true. AppSync doesn't add fields. You don't mention using Amplify CLI but from the `@model` and `createdAt`/`updatedAt` fields it appears you are. It's Amplify CLI that is adding those fields.

I know lot of people use "AppSync" and "Amplify" interchangeably so it's become confusing but they're different products.

If you want those fields to be nullable then you should raise an issue with Amplify CLI.

If you want AppSync to return all fields requested then there is already an issue as I mentioned.

**Comment by gsans** [comment]
I think by now is clear what is the issue and what are the possible solutions. The context is that the current implementation leads to bad user experience and that needs to be fixed.
